File: LeetCode/Easy/001_Two_Sum/solution.py
"""
Problem: Two Sum
Platform: LeetCode
Difficulty: Easy
Time Complexity: O(n)
Space Complexity: O(n)
"""

from typing import List

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        """
        Find two indices where nums[i] + nums[j] = target
        
        Args:
            nums: List of integers
            target: Target sum
            
        Returns:
            List containing two indices [i, j]
        """
        # Hash map to store number and its index
        num_map = {}
        
        # Iterate through the array
        for i, num in enumerate(nums):
            # Calculate the complement
            complement = target - num
            
            # Check if complement exists in map
            if complement in num_map:
                # Found the pair, return indices
                return [num_map[complement], i]
            
            # Add current number to map
            num_map[num] = i
        
        # No solution found - this line ensures function returns a value
        # but should never execute as problem guarantees exactly one solution exists
        return []

# Example Usage
if __name__ == "__main__":
    sol = Solution()
    
    # Test Case 1
    nums1 = [2, 7, 11, 15]
    target1 = 9
    print(f"Input: nums = {nums1}, target = {target1}")
    print(f"Output: {sol.twoSum(nums1, target1)}")  # [0, 1]
    
    # Test Case 2
    nums2 = [3, 2, 4]
    target2 = 6
    print(f"\nInput: nums = {nums2}, target = {target2}")
    print(f"Output: {sol.twoSum(nums2, target2)}")  # [1, 2]
    
    # Test Case 3
    nums3 = [3, 3]
    target3 = 6
    print(f"\nInput: nums = {nums3}, target = {target3}")
    print(f"Output: {sol.twoSum(nums3, target3)}")  # [0, 1]

File: LeetCode/Easy/001_Two_Sum/solution.cpp
/**
 * Problem: Two Sum
 * Platform: LeetCode
 * Difficulty: Easy
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */

#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        // Hash map to store number and its index
        unordered_map<int, int> numMap;
        
        // Iterate through the array
        for (int i = 0; i < nums.size(); i++) {
            // Calculate the complement
            int complement = target - nums[i];
            
            // Check if complement exists in map
            if (numMap.find(complement) != numMap.end()) {
                // Found the pair, return indices
                return {numMap[complement], i};
            }
            
            // Add current number to map
            numMap[nums[i]] = i;
        }
        
        // No solution found (shouldn't happen based on problem constraints)
        return {};
    }
};

/**
 * Example Usage:
 * 
 * Solution sol;
 * vector<int> nums = {2, 7, 11, 15};
 * int target = 9;
 * vector<int> result = sol.twoSum(nums, target);
 * // result = {0, 1}
 */
